x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/CachingUsernamePasswordRealm.java

```diff
@@ -263,16 +263,21 @@ private void lookupWithCache(String username, ActionListener<User> listener) {
 
     protected abstract void doLookupUser(String username, ActionListener<User> listener);
 
-    private static class UserWithHash {
-        final User user;
-        final char[] hash;
+    private static class CachedResult {
+        private final AuthenticationResult authenticationResult;
+        private final User user;
+        private final char[] hash;
 
-        UserWithHash(User user, SecureString password, Hasher hasher) {
-            this.user = Objects.requireNonNull(user);
+        private CachedResult(AuthenticationResult result, Hasher hasher, @Nullable User user, @Nullable SecureString password) {
```

Comment:
Given the `AuthenticationResult` has the `User`, do we need to pass both in to the constructor?
Having only the `result` would remove the need to check that they're consistent (both null/not-null)